### ISSUE TITLE
arrgghh milk broken?

### ISSUE THREAD

**Issue by drbitboy** [milk-org/milk#13]
It looks like a few of the submodules and/or other pieces of milk have changed recently (e.g. the signature for delete_image), can anyone give me a road map to get things back together?

Jared and I have a fork of milk/dev called resurrector, and I think some items are out of sync with the submodules?  maybe?  e.g.

```
‘const struct pybind11::buffer_info’ has no member named ‘item_type_is_equivalent_to’
```

Thanks, Brian

**Comment by DasVinch** [comment]
Hey @drbitboy,

Yes, Oli's been doing a lot of changes.

As far as I'm seeing, the trifecta milk @ dev branch + `git submodule update` + `./fetch_cacao_dev.sh` compiles fine.

I'm posting this with:
- milk ae675570219559ee7703c595abeb447f4ae35f8c
- ISIO https://github.com/milk-org/ImageStreamIO/commit/0c5944820fe3decf6aa9f1d47611b0b938b3bdf6
- CACAO https://github.com/cacao-org/cacao/commit/ae3e2bee261e11188c38d5d017596d2cedb560d3

**Comment by DasVinch** [comment]
Edit: this is in the python_module extension (of ISIO), due to a change by @a-sevin to fix int64 types in the past week.

This is actually an extremely recent addition to pybind? It shows at july 14th 2023 in the changelog.
https://pybind11.readthedocs.io/en/stable/changelog.html

Recommended fixes: update pybind11-dev on your system, or compile without -Dbuild_python_module=ON

**Comment by drbitboy** [comment]
What version of Mambaforge/Miniforge is Oli using?

**Comment by drbitboy** [comment]
I installed the latest version of conda (```Mambaforge-Linux-x86_64.sh```) and I did not get ```item_type_is_equivalent_to``` in ```/opt/conda/include/pybind11/buffer_info.h```

arrgghh

**Comment by drbitboy** [comment]
> Hey @drbitboy,
> 
> Yes, Oli's been doing a lot of changes.
> 
> As far as I'm seeing, the trifecta milk @ dev branch + `git submodule update` + `./fetch_cacao_dev.sh` compiles fine.
> 
> I'm posting this with:
> 
>     * milk [ae67557](https://github.com/milk-org/milk/commit/ae675570219559ee7703c595abeb447f4ae35f8c)
> 
>     * ISIO [milk-org/ImageStreamIO@0c59448](https://github.com/milk-org/ImageStreamIO/commit/0c5944820fe3decf6aa9f1d47611b0b938b3bdf6)
> 
>     * CACAO [cacao-org/cacao@ae3e2be](https://github.com/cacao-org/cacao/commit/ae3e2bee261e11188c38d5d017596d2cedb560d3)

so what is the conda/python environment that makes that trifecta compile?

Is Olivier seeing this thread?

**Comment by DasVinch** [comment]
So actually, for me cmake is only detecting the pybind I have installed with apt... and the latest pybind11-dev on Ubuntu 20.04 seems to cut it.

The pybind from my conda is used (on imagestreamIO) when i compile pymilk through a pip install.

Will give version numbers in a little.

**Comment by DasVinch** [comment]
- pybind11 over conda: 2.11.1
- pybind11-dev over APT: 2.4.3-2build2
I needed to install the latter in other to achieve a succeful python_build. But per the changelog `item_type_is_equivalent_to` is only available in the former. Confusion is real...

**Comment by joseph-long** [comment]
In MagAO-X we pin versions of all our Python dependencies, which in this case includes pybind11 from conda-forge, so it was necessary to explicitly `mamba uninstall pybind11` and then `mamba install pybind11`. 

Now:

```
[ 99%] Linking C shared library libcacaocomputeCalib.so
/usr/bin/ld: cannot find -llapacke
collect2: error: ld returned 1 exit status
make[2]: *** [plugins/cacao-src/computeCalib/CMakeFiles/cacaocomputeCalib.dir/build.make:304: plugins/cacao-src/computeCalib/libcacaocomputeCalib.so] Error 1
```

We jump through some hoops to build with MKL, which I thought meant we don't need `-llapacke` to get the lapacke symbols—since MKL provides them. Can we patch the CMakeLists to omit it when MKL is used?

**Comment by joseph-long** [comment]
It looks like this works:

```diff
diff --git a/computeCalib/CMakeLists.txt b/computeCalib/CMakeLists.txt
index 15b1646b..a6df65b4 100644
--- a/computeCalib/CMakeLists.txt
+++ b/computeCalib/CMakeLists.txt
@@ -44,10 +44,11 @@ set(LINKLIBS
        CLIcore
        milkinfo
        cacaoAOloopControl
-       lapacke
 )

-
+if(NOT MKL_FOUND)
+list(APPEND LINKLIBS lapacke)
+endif()
```

**Comment by drbitboy** [comment]
> In MagAO-X we pin versions of all our Python dependencies, which in this case includes pybind11 from conda-forge, so it was necessary to explicitly `mamba uninstall pybind11` and then `mamba install pybind11`.

Nope.  There is no package pybind11, it is pybind11-abi, and when we uninstall pybind11-abi, then it also uninstalls mamba, so the second command ```mamba install ...``` fails.

**Comment by joseph-long** [comment]
I updated to a newer version of the mamba installer as well, which may have changed the way they treat special dependencies of the mamba solver

**Comment by drbitboy** [comment]
I am working on automated provisioning of a MagAO-X environment, so minimizing manual interventions is a design constraint.  I found a [mamba env export] YAML file that is being used, with version numbers; maybe that is the problem ...

Basically, we are working on two different problems:  you are updating code while you simultaneously update the tools that code uses (e.g. you updated mamba); we are using only the updated code, which breaks with the older tools we are using.

A more comprehensive test environment would be nice, but MagAO-X is one of MILK's and CACAO's many clients, and we cannot hand off responsibility for the MagAO-X build to someone else, so I guess this kind of circus is inevitable.

I'll close this issue; let me know if you want to discuss it some more.

**Comment by sanfords** [comment]
I think we need _both_ the tools as well as the code under CM in this repo.  For the tools, we need automatable ways to get them.

**Comment by joseph-long** [comment]
I agree, this seems off topic for the milk repo. @DasVinch identified the version of pybind11 required and I identified the places in the MagAO-X repo that need to be updated to match. When i have finished setting up the new OS install on the testbed I will push those updates to the GitHub copy. If you would like to open a tracking issue on magao-x/MagAOX i will update it when the changes land.

**Comment by DasVinch** [comment]
Issue already closed, but just letting you know I've reproduced on a scexao server:
```
(base) scexao@scexao6:✔0:~$ conda list | grep pybind
pybind11                  2.8.1                    pypi_0    pypi
(base) scexao@scexao6:✔0:~$ apt show pybind11-dev 
Package: pybind11-dev
Version: 2.4.3-2build2
```

Somehow, having the APT package resolves my CMAKE resolving issues, but the version actually used is the conda one... we might totally be talking two different problems.
TLDR, with conda pybind < 2.11, I can't compile either.

**Comment by drbitboy** [comment]
I updated the resurrector branches to work with multipass.  This now works for me, from scratch:
```
===================
=== musudo.bash ===
===================

    #!/usr/bin/env bash

    ### Probably add some of this to ensure_dirs_and_perms.sh

    apt-get update                                        ### Short wait
    mkdir -p /usr/local/share/man/man{1,2,3,4,5,6,7,8,9}
    mkdir -p /opt/MagAOX
    chown :ubuntu /opt/MagAOX
    chmod u+sX,g+wsX /opt/MagAOX
    echo export MAGAOX_ROLE=vm | tee    /etc/profile.d/magaox_role.sh
    exit 0

===============
=== On host ===
===============

    mu launch -n five -c 5 -d 20GiB -m 8.0GiB 22.04
    mu exec five -- touch .hushlogin
    mu transfer tmp/musudo.bash five:musudo.bash
    #                                               ### Extra carcich host steps
    mu shell five


=======================
=== On guest (five) ===
=======================

    sudo ./musudo.bash
    #                                               ### Extra carcich guest steps
    mkdir githubalt
    cd githubalt/
    git clone   -b resurrector      https://github.com/drbitboy/MagAOX.git
    git clone   -b resurrector      https://github.com/drbitboy/magao-x-config.git config
    git clone   -b dev-no-lapacke   https://github.com/milk-org/milk.git
    cd MagAOX/setup
    bash -lx provision.sh


================================
=== Extra carcich host steps ===
================================
    mu transfer ~/tmp/ssh.tar five:.ssh/ssh.tar
    mu exec five -- tar xvf .ssh/ssh.tar -C .ssh/


=================================
=== Extra carcich guest steps ===
=================================

    touch .ssh/known_hosts
    chmod 600 .ssh/known_hosts
    eval $(ssh-agent)
    ssh-add .ssh/github_id_rsa
```
